### ISSUE TITLE
Not able to connect with Remote DB

### ISSUE THREAD

**Issue by amanattrish** [horilla-opensource/horilla#295]
# Bug Report

## Stuck in infinite loop while running migrations command

I'm trying to use MySQL over AWS RDS. So I removed the DATABASE_URL and added the NAME, HOST, USERNAME, PASSWORD and PORT in `horilla/settings.py` file. Then while running migration command `python3 manage.py migrations` I get following

<img width="898" alt="image" src="https://github.com/user-attachments/assets/e18474ce-af83-4da3-8800-efab458cd4aa">

## Steps to Reproduce

1. In docker-compose.yml file, comment out the pgsql part and dependency of server in pgsql
2. Add mysqlclient as given in database setup docs - https://www.horilla.com/docs/database-setup/
3. run the `docker compose up` command

## Expected Behavior

It should run without any problem

## Actual Behavior

Getting `Execution of job "block_unblock_disciplinary (trigger: interval[0:00:10], next run at: 2024-08-25 17:35:45 IST)" skipped: maximum number of running instances reached (1)`

## Screenshots

_[If applicable, provide screenshots illustrating the issue.]_

## Environment

- **Django Version:** _[Django version you're using, e.g., 3.2.1]_
- **Python Version:** _[Python version, e.g., 3.9]_
- **Operating System:** _[Your OS, e.g., Windows 10, Ubuntu 20.04]_
- **Browser:** _[If applicable, specify the browser and version you're using.]_

## Additional Information

_[Any other relevant information, logs, error messages, or context that might help in understanding and fixing the issue.]_

## Possible Solution

_[If you have any ideas or suggestions on how to fix the issue, feel free to provide them here. This is optional.]_

## Labels

_[If your project uses labels, suggest any labels that might apply to this issue, such as 'bug', 'needs investigation', etc.]_

## Priority

_[Specify the priority level for this issue, such as 'high', 'medium', 'low', etc.]_

## Assignees

_[If you want to suggest an assignee or tag a specific person to look into this issue, mention their GitHub username here.]_

## Related Issues

_[If there are any related issues or pull requests, mention them here. This is optional.]_

**Note:** _Remember to search through existing issues before submitting a new one to ensure that the issue hasn't been reported already. Provide as much information as possible to help the maintainers understand and address the problem effectively._

**Comment by amanattrish** [comment]
If we run it long enough, it throws following error

`Job "block_unblock_disciplinary (trigger: interval[0:00:10], next run at: 2024-08-25 18:06:48 IST)" raised an exception`
`django.db.utils.ProgrammingError: (1146, "Table 'horilla.employee_disciplinaryaction' doesn't exist")`

**Comment by amanattrish** [comment]
Update:

The same problem is happening when I try to connect with PgSQL. But when I try to connect spin a pgsql docker container using docker-compose, it works fine.

**Comment by amanattrish** [comment]
Update:

It all seems like db connection error, I tried same credentials in Dbeaver, TablePlus, Terminal and other connectivity softwares, connection to database is fine, its just horilla code does not connect.

Is there a way I can test connection from code?

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
For time being the below warning can be neglected.
`
Execution of job "block_unblock_disciplinary (trigger: interval[0:00:10], next run at: 2024-08-25 17:35:45 IST)" skipped: maximum number of running instances reached (1)
`
It's coming from an scheduler job which is running in the background. 
Once we comment out the scheduler, that warning is missed. But the connection to RDS is not getting executed is what I think the issue in your case.
Can you test whether the connection to RDS is not being blocked? Can you confirm with the security group if the requests from your server IP is accepted ?

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
Hi @horilla-opensource 

Thanks for reply. Yes it is Db connection issue.

I tested the connection using 1. command line, 2. Different Django project, 3. Dbeaver (a DB tool) and 4. TablePlus (a DB tool). The connection to AWS RDS PostgreSQL as well as AWS RDS MySQL are completely fine.

Moreover, I see the localhost db is also connecting without any problem. But its just remote DB which is unable to connect.

Local DB configs
```python
DATABASES = {
    "default": {
        "ENGINE": 'django.db.backends.postgresql',
        "NAME": 'my_db_name',
        "USER": 'my_user',
        "PASSWORD": 'my_password',
        "HOST": 'localhost',
        "PORT": '5432',
    }
}
```

Remote DB configs

```python
DATABASES = {
    "default": {
        "ENGINE": 'django.db.backends.postgresql',
        "NAME": 'my_db_name',
        "USER": 'my_user',
        "PASSWORD": 'my_password',
        "HOST": 'xxxxx.xxxxxx.us-east-2.rds.amazonaws.com',
        "PORT": '5432',
    }
}
```

**Comment by amanattrish** [comment]
I even tried `django.db.backends.postgresql_psycopg2`

as mentioned in here - https://stackoverflow.com/questions/57984929/django-app-is-not-connecting-with-remote-database.

For your kind information, I'm using Macbook air  with M2 chip.

**Comment by horilla-opensource** [comment]
@BhuviTheDataGuy Do you have any suggestions on what can be the issue causing this to fail ?
Any kind of help or suggestion is much appreciated.

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
Any updates @horilla-opensource @BhuviTheDataGuy?

**Comment by BhuviTheDataGuy** [comment]
@amanattrish 

The error you mentioned in the previous comment is 
```
django.db.utils.ProgrammingError: (1146, "Table 'horilla.employee_disciplinaryaction' doesn't exist")
```
Is it coming from the RDS database(your remote DB)? 

If yes, then I don't think it is a connection error, After connecting to the database, it was not able to find the `employee_disciplinaryaction` table. Here the database connection was fine.

Could you please help me to understand this,
The connection to AWS RDS PostgreSQL as well as AWS RDS MySQL are completely fine, But its just remote DB which is unable to connect. -- Are you referring these AWS RDS PostgreSQL/MySQL as the remote DB or something else?


What else you are seeing in the error log related to database?

**Comment by amanattrish** [comment]
Hi @BhuviTheDataGuy 
Thanks for your reply.

1. The error came because migrations did not run at all and this error comes from disciplinary-action job. I suppose it is responsibility of migrations to create all those tables in DB. Strange part is, in local DB these errors don't come.
```
django.db.utils.ProgrammingError: (1146, "Table 'horilla.employee_disciplinaryaction' doesn't exist")
```

2. Yes I'm referring these AWS RDS PostgreSQL/MySQL as the remote DB. If you want I can share the PgSQL creds of a public DB I have so that you can also test from your end.

**Comment by BhuviTheDataGuy** [comment]
Thanks for the clarifications. 

Are you able to see anything related to DB connection error in the log file? That would help to identify the root cause of this issue. 

If possible, could you please create the container again and share the compose log file.

**Comment by amanattrish** [comment]
Hi @BhuviTheDataGuy 

Its this 

Terminal Screenshot
<img width="1061" alt="image" src="https://github.com/user-attachments/assets/a2bf42dc-41fc-4f03-8700-fe914e580786">

Docker desktop screenshot
<img width="971" alt="image" src="https://github.com/user-attachments/assets/6174e469-07d4-4346-8717-148cfcabf1f0">

[logs.txt](https://github.com/user-attachments/files/16792604/logs.txt)


The job is not able to run because of connection I guess, if I comment out the job code, it does not do anything.

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
To avoid the warning , please comment out the below line in employee module 
[https://github.com/horilla-opensource/horilla/blob/6f6a19e0ba59821f2acf6ae58c98f4190a3de108/employee/scheduler.py#L137](https://github.com/horilla-opensource/horilla/blob/6f6a19e0ba59821f2acf6ae58c98f4190a3de108/employee/scheduler.py#L137) and try again.

**Comment by amanattrish** [comment]
its just stuck here 
<img width="826" alt="image" src="https://github.com/user-attachments/assets/a82e58eb-f686-43a8-8505-8bf0e8c7d727">
<img width="982" alt="image" src="https://github.com/user-attachments/assets/b031f7d5-3bc2-4584-ae07-c1e8d22a7ca4">

For your kind information, `Running makemigrations...` is the log I added in`entrypoint.sh` file
<img width="886" alt="image" src="https://github.com/user-attachments/assets/dcf39179-0341-4ebf-bcb6-a5274ee51ec2">

**Comment by amanattrish** [comment]
@horilla-opensource 

Can somebody from your team try to connect to PgSQL or MySQL in RDS? I have asked my friend to do the same, it is a problem in his machine as well.

CC: @BhuviTheDataGuy

**Comment by amanattrish** [comment]
Any updates @BhuviTheDataGuy @horilla-opensource ?

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
We were able to reproduce the issue at our end and the team is working on it .
Will provide you an update once we were able to find a solution for it.

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
Thanks @horilla-opensource for the updates.

**Comment by amanattrish** [comment]
@horilla-opensource 
I know the team might be busy, just wanted to know if there is any time expectation to have it resolved?

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
We get your concern on the delay. The team is working on the issue and will be able to provide you the solution mostly by EOD tomorrow.

With Regards,
**Team Horilla**

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
We apologize for the delay being caused. We still find some difficulty in identifying the issue with the RDS connection.
We'll update with once we have the solution or walk around for the same.

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
No problem @horilla-opensource 

Thanks for update. I really appreciate it.

**Comment by amanattrish** [comment]
Hi @horilla-opensource 

Any luck in finding out a resolution or walk around for rds db connection?

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
Can you implement the system currently with a on premise database till we find a solution to this? Once fixed, we can take the backup from the local and update it to the RDS.

**Comment by amanattrish** [comment]
Sure Thanks

**Comment by centaurusgod** [comment]
> Hi @BhuviTheDataGuy
> 
> Its this
> 
> Terminal Screenshot <img alt="image" width="1061" src="https://private-user-images.githubusercontent.com/53205776/362547351-a2bf42dc-41fc-4f03-8700-fe914e580786.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyNzYwMzYsIm5iZiI6MTcyNTI3NTczNiwicGF0aCI6Ii81MzIwNTc3Ni8zNjI1NDczNTEtYTJiZjQyZGMtNDFmYy00ZjAzLTg3MDAtZmU5MTRlNTgwNzg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDExMTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxNjE5N2Y4NjhjYWI0MDVmZDBkYzYwYWNmMWQ4YzYxMTliMjBlMDQ4OTBmYjkwNDg5MjFjNWJjMDVmYzNkMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nK-75E0rNzG7_zDZxG1eui7FT_G8l24UVz1S_W_qNSw">
> 
> Docker desktop screenshot <img alt="image" width="971" src="https://private-user-images.githubusercontent.com/53205776/362547471-6174e469-07d4-4346-8717-148cfcabf1f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyNzYwMzYsIm5iZiI6MTcyNTI3NTczNiwicGF0aCI6Ii81MzIwNTc3Ni8zNjI1NDc0NzEtNjE3NGU0NjktMDdkNC00MzQ2LTg3MTctMTQ4Y2ZjYWJmMWYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDExMTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkYTE2ODI5YWU1OGNmM2FhNGQ4ODE3MzJlOTEzMTVlMWQ2Mzg0ZDRjNDBjODdhNjBkNzM5YmU1NGY0YjU1ODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.K_kHQ0cdR-sZ4Uh5ivjWFpaA35pyAKPjEBO5hzQAGG0">
> 
> [logs.txt](https://github.com/user-attachments/files/16792604/logs.txt)
> 
> The job is not able to run because of connection I guess, if I comment out the job code, it does not do anything.

Don't know if this information will help but..
I had encountered the exact same issue (block_unblock_disciplinary) under these two conditions
a. Using MySQL server (Fixed by switching to MariaDB)
b. Nginx and Gunicorn were running in the background while I was trying to access the Django server through the public IP. 
`python manage.py runserver 0.0.0.0:8000` (No errors after stopping those)


This issue was absent when running Django in local server and only occurred after I migrated the application to Lightsail instance.

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
Were you test the method suggested by @centaurusgod ?

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
Hi @horilla-opensource 

1. We have MySQL and PgSQL instances. Using MariaDB just for 1 more thing may not be best solution for us.
2. We don' have Nginx and Gunicorn running in background. Unlike @centaurusgod we are getting error in local server and AWS ECS both.

There is one thing I wanted to point out, if I create new Django project and use the same DB creds (be it MySQL, PgSQL), the db connection works perfectly fine be it local server or AWS ECS.

**Comment by amanattrish** [comment]
Hi @horilla-opensource 

Any luck in finding resolution?

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
We are not able to correctly identify what's causing the connection problem. 
The connection was getting refused all the time, and eventually the ip getting blocked by AWS after regular tries (both in local and aws).
We have allocated a separate team for looking into this. They are working on this now.
Apologize for the delay.

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
Thanks for update. Really appreciate it.

**Comment by amanattrish** [comment]
@horilla-opensource did the team find anything.

**Comment by amanattrish** [comment]
Hi @horilla-opensource 
Any luck on finding root cause of the problem?

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,

We’ve been actively investigating the connection issue with the remote database in Horilla and have been working on it for some time. While we’ve made progress, we haven’t yet pinpointed the exact root cause of the problem.

We’re continuing to explore potential solutions and would appreciate your understanding as we work through this. Rest assured, this is a priority for us, and we will update you as soon as we have a concrete resolution.

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
Thank you so much

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,

We are now able to successfully connect our remote MySQL and Postgres databases from AWS with Horilla. 

There was no issue with the system. The connection issue was from the AWS RDS settings. Please follow the following steps with the creation of database and try from your side. (Also refer the the youtube video : [https://www.youtube.com/watch?v=HGOrsBzQrq0](https://www.youtube.com/watch?v=HGOrsBzQrq0) )

1. Choose _Standard Create_ and your desired database(choosing postgres as reference here)
<img width="1139" alt="Screenshot 2024-09-17 at 3 48 53 PM" src="https://github.com/user-attachments/assets/6fc96786-d589-4e97-ab8a-eda2c3941345">

2. Choose your desired template
<img width="1140" alt="Screenshot 2024-09-17 at 3 49 18 PM" src="https://github.com/user-attachments/assets/4b67d56e-068f-4e88-9a84-96f0f81d82f7">

3. Inside the _Settings_, add a name for the _DB instance identifier_ and add the name for the user (by default, it'll be _postgres_, this user will be used inside the Horilla database configuration settings)
<img width="1139" alt="Screenshot 2024-09-17 at 3 49 50 PM" src="https://github.com/user-attachments/assets/e7e925f4-3053-4c38-88a4-8d81e085ac29">

4. Select _Self managed_ and add your master password
<img width="1140" alt="Screenshot 2024-09-17 at 3 50 13 PM" src="https://github.com/user-attachments/assets/f2c54be9-119e-4651-8f14-4d9a4b97b71a">

5.
<img width="1140" alt="Screenshot 2024-09-17 at 3 50 43 PM" src="https://github.com/user-attachments/assets/ea6a456d-1879-46fa-9c61-4f44ce6e6bd2">

6. Don't connect the RDS to any EC2 instance, and enable the _Public access_
<img width="1141" alt="Screenshot 2024-09-17 at 3 50 54 PM" src="https://github.com/user-attachments/assets/1ea7fbca-7005-42fc-8db4-17fb5f180edf">

7. Choose existing security group 
<img width="1141" alt="Screenshot 2024-09-17 at 3 51 05 PM" src="https://github.com/user-attachments/assets/753c68c8-4672-4c7d-827a-2ae5c4da0a65">

8.
<img width="1141" alt="Screenshot 2024-09-17 at 3 51 21 PM" src="https://github.com/user-attachments/assets/b302ea44-f9c8-4af8-9df9-a1682872d840">

9. Create a new database for Horilla. Add a name for the database. This name will be used inside the settings.
<img width="1141" alt="Screenshot 2024-09-17 at 3 51 47 PM" src="https://github.com/user-attachments/assets/b79516ac-1f69-4eb0-8c95-168d67ca2546">

10. Once the RDS is successfully created, open its dashboard
<img width="1142" alt="Screenshot 2024-09-17 at 3 57 40 PM" src="https://github.com/user-attachments/assets/48483112-8523-4973-ba12-5aa5f2c3d9d9">

11. Inside the RDS dashboard, click on the security group
<img width="1133" alt="Screenshot 2024-09-17 at 3 57 50 PM" src="https://github.com/user-attachments/assets/33c23f10-14ae-4246-a0e7-3646905d0fe8">

12. Select the security group
<img width="1180" alt="Screenshot 2024-09-17 at 3 57 58 PM" src="https://github.com/user-attachments/assets/16302cd9-e785-4d1b-a7c4-eadbb16b2796">

13. Edit the outbound rules
<img width="1174" alt="Screenshot 2024-09-17 at 3 58 05 PM" src="https://github.com/user-attachments/assets/f6e25654-8aba-44c8-8f20-b1fd96a4a77d">

14. Select type as _Postgres_ and Destination as _Anywhere-IPv4_ and save the settings.
<img width="1431" alt="Screenshot 2024-09-17 at 3 58 34 PM" src="https://github.com/user-attachments/assets/528e23e4-f3cb-45d9-b500-e9127a56afc6">

15. Similarly edit the inbound rules also
<img width="1191" alt="Screenshot 2024-09-17 at 4 03 30 PM" src="https://github.com/user-attachments/assets/e4a181ee-69aa-469a-837e-a73657bb9b10">

16. Please remove all other inbound or outbound rules other than the 5432 port rule (for initial testing)
<img width="1423" alt="Screenshot 2024-09-17 at 4 04 30 PM" src="https://github.com/user-attachments/assets/9487c9c1-d1a1-44ca-af48-9e3c2cad8562">

17. Copy the endpoint value.
<img width="1147" alt="Screenshot 2024-09-17 at 4 06 28 PM" src="https://github.com/user-attachments/assets/c69fc4fe-a184-4e8f-8956-df24c32f0b60">

18. Edit the settings.py file with the new configuration values.
<img width="559" alt="Screenshot 2024-09-17 at 4 08 11 PM" src="https://github.com/user-attachments/assets/ab681927-dd60-4413-841a-0b1de5c060ba">


Please let me know if you face any issues.

With Regards,
**Team Horilla**

**Comment by amanattrish** [comment]
Hi @horilla-opensource 

I followed exactly same, still same issue.
Is there an option for paid support from your side?

**Comment by amanattrish** [comment]
I have one followup question as well, if it is from AWS, why does it connects to new Django project without any problem?

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
Please connect at support@horilla.com for support.

**Comment by horilla-opensource** [comment]
Hi @amanattrish ,
We are closing this issue due to inactivity. It seems that there hasn't been any recent activity or discussion, and we believe it may have been resolved or is no longer relevant. 
If you feel this issue is still important and should be addressed, please feel free to reopen it or create a new issue with updated details. 

With Regards,
**Team Horilla**
